packages/x-components/src/composables/use-scroll.ts

```diff
@@ -0,0 +1,309 @@
+    return !hasScrollReachedStart.value && props.distanceToBottom > distanceToEnd.value;
+  });
+
+  onMounted(() => {
+    nextTick().then(() => {
+      if (!scrollEl) {
+        // TODO Replace with Empathy's logger
+        // eslint-disable-next-line no-console
+        console.warn(
+          '[ScrollMixin]',
+          'Components using this mixin must set `this.$el` to the HTML node that is scrolling.'
+        );
+      } else {
+        storeScrollData();
+      }
+    });
+  });
+
+  /**
+   * Change the isClientHeightChanging flag when the property clientHeight is changing and
+   * calls throttleledCall method.
+   *
+   * @internal
+   */
+  watch(
+    clientHeight,
+    () => {
+      isClientHeightChanging.value = true;
+
+      throtteledCall();
+    },
+    { immediate: true }
+  );
+
+  /**
+   * Emits the `scroll` event.
+   *
+   * @param _newScrollPosition - The new position of scroll.
+   * @param oldScrollPosition - The old position of scroll.
+   * @internal
+   */
+  watch(currentPosition, (_newScrollPosition: number, oldScrollPosition: number) => {
+    context.emit('scroll', currentPosition.value);
+    previousPosition.value = oldScrollPosition;
+  });
+
+  /**
+   * Sets direction of scroll based in {@link ScrollDirection} when the current position
+   * has changed.
+   *
+   * @internal
+   */
+  watch(currentPosition, () => {
+    if (!isClientHeightChanging.value && currentPosition.value !== previousPosition.value) {
+      scrollDirection.value = currentPosition.value > previousPosition.value ? 'DOWN' : 'UP';
+    }
+  });
+
+  /**
+   * Emits the 'scroll:almost-at-end' event when the user is about to reach to end.
+   *
+   * @param isScrollAlmostAtEnd - For knowing if the user is about to reach to end.
+   * @internal
+   */
+  watch(hasScrollReachedStart, (isScrollAtStart: boolean) => {
+    context.emit('scroll:at-start', isScrollAtStart);
+  });
+
+  /**
+   * Sets direction of scroll based in {@link ScrollDirection} when the current position
+   * has changed.
+   *
+   * @internal
+   */
+  watch(hasScrollAlmostReachedEnd, (isScrollAlmostAtEnd: boolean) => {
+    context.emit('scroll:almost-at-end', isScrollAlmostAtEnd);
+  });
+
+  /**
+   * Emits the 'scroll:at-end' event when the user reaches the end.
+   *
+   * @param isScrollAtEnd - For knowing if the user reaches at end.
+   * @internal
+   */
+  watch(hasScrollReachedEnd, (isScrollAtEnd: boolean) => {
+    context.emit('scroll:at-end', isScrollAtEnd);
+  });
+
+  /**
+   * Emits the `scroll:direction-change` event when the scrolling direction has changed.
+   *
+   * @param direction - The new direction of scroll.
+   * @internal
+   */
+  watch(scrollDirection, (direction: ScrollDirection) => {
+    if (!isClientHeightChanging.value) {
+      context.emit('scroll:direction-change', direction);
+    }
+  });
+
+  /**
+   * Resets the scroll position.
+   *
+   * @internal
+   */
+  const $x = use$x();
```

Comment:
If you guarantee the `resetOn` prop as always array, you can loop it better. Something like: 
```TS
  const $x = use$x();
  const resetOnEvents = isArray(props.resetOn) ? props.resetOn : [props.resetOn];
  resetOnEvents.forEach(event =>
    $x.on(event, false).subscribe(() =>
      nextTick().then(() => {
        if (props.resetOnChange) {
          scrollEl.value?.scrollTo({ top: 0 });
        }
      })
    )
  );
```